python/pyspark/sql/connect/client/retries.py

```diff
@@ -0,0 +1,228 @@
+        if wait_time >= self.policy.min_jitter_threshold:
+            wait_time += random.uniform(0, self.policy.jitter)
+
+        # Round to whole number of milliseconds
+        return int(wait_time)
+
+
+class AttemptManager:
+    """
+    Simple ContextManager that is used to capture the exception thrown inside the context.
+    """
+
+    def __init__(self, retrying: "Retrying") -> None:
+        self._retrying = retrying
+
+    def __enter__(self) -> None:
+        pass
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exception: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> Optional[bool]:
+        if isinstance(exception, BaseException):
+            # Swallow the exception.
+            if self._retrying.accept_exception(exception):
+                return True
+            # Bubble up the exception.
+            return False
+        else:
+            self._retrying.accept_succeeded()
+            return None
+
+
+class Retrying:
+    """
+    This class is a point of entry into the retry logic.
+    The class accepts a list of retry policies and applies them in given order.
+    The first policy accepting an exception will be used.
+    """
+
+    def __init__(
+        self,
+        policies: typing.Union[RetryPolicy, typing.Iterable[RetryPolicy]],
+        sleep: Callable[[float], None] = time.sleep,
+    ) -> None:
+        if isinstance(policies, RetryPolicy):
+            policies = [policies]
+        self._policies: List[RetryPolicyState] = [policy.to_state() for policy in policies]
+        self._sleep = sleep
+
+        self._exception: Optional[BaseException] = None
+        self._done = False
+
+    def can_retry(self, exception: BaseException) -> bool:
+        return any(policy.can_retry(exception) for policy in self._policies)
+
+    def accept_exception(self, exception: BaseException) -> bool:
+        if self.can_retry(exception):
+            self._exception = exception
+            return True
+        return False
+
+    def accept_succeeded(self):
+        self._done = True
+
+    def _last_exception(self) -> BaseException:
+        if self._exception is None:
+            raise RuntimeError("No active exception")
+        return self._exception
+
+    def _wait(self):
+        exception = self._last_exception()
+
+        # Attempt to find a policy to wait with
+        for policy in self._policies:
+            if not policy.can_retry(exception):
+                continue
+
+            wait_time = policy.next_attempt()
+            if wait_time is not None:
+                logger.debug(
+                    f"Got error: {exception}. "
+                    + f"Will retry after {wait_time} ms (policy: {policy.name})"
+                )
+
+                self._sleep(wait_time / 1000)
+                return
+
+        # Exceeded retries
+        logger.debug(f"Given up on retrying. error: {exception}")
+        raise exception
```

Comment:
from offline discussion / unrelated to the core "support multiple retry policies":
maybe it would be nice to throw a specific exception type like "RetryExceeded", with all the previously retried exceptions wrapped inside? A user is unlikely to catch on exceptions that are normally retried, but they may want to catch one special exception type that will give them information what transient errors persisted and for how long, over how many retries etc.

Reply:
That seems sensible, we can do that